### PR TITLE
cross-platform pex/whl file with cryptography installed

### PR DESCRIPTION
**Pull request by lyw07**
## Summary

Creates a cross-platform pex/whl file which include the c extension cryptography. 

- Have tests been written for the new code?
   - Tested on Linux x86_64 with python 2.7, 3.4, 3.5, 3.6 and worked, 
   - Tested on win32 and win-amd64 with python 2.7, 3.4, 3.5, 3.6 (although 3.4 doesn’t work because of syntax errors from cherrypy and mathjax) and worked, 
   - Tested on Mac OS X 10.11 with python 2.7, 3.4, 3.5, 3.6 and worked.

- New dependencies (if any) added to requirements file
   - Added a new requirement file named cext.txt inside the requirements folder. It includes the name of the package with the version. The install_cexts.py needs to read both name and version of the package to successfully install the correct package.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -118,3 +124,8 @@ dockerenvbuild: writeversion
 
 dockerenvdist: writeversion
 	docker run -v $$PWD/dist:/kolibridist learningequality/kolibri:$$(cat kolibri/VERSION)
+
+kolibripippex:
+	git clone https://github.com/lyw07/pip.git
```
Comment (jamalex) [inline Makefile]: Let's move this repo onto `learningequality`

File: install_cexts.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        subprocess.call(['pip', 'install', 'bs4'])
+
+
+"""
+Distribute packages into folders according to the platform.
+"""
+def get_path_with_arch(platform, path):
+    # linux_i686 is not supported
+    if 'i686' in platform:
+        print ('Kolibri is not supported on this platform. Skipped.\n')
```
Comment (jamalex) [inline install_cexts.py]: Since cryptography is an optional dependency, I don't know we want to say "Kolibi is not supported". Maybe "C Extensions are not supported on this platform".

File: kolibri/utils/cli.py [hunk 3]
```diff
@@ -14,10 +15,44 @@
...
+if 'linux' in platform and int(python_version[2:]) < 33:
+    dirname = os.path.join(dirname, 'linux')
+    # encode with ucs2
+    if sys.maxunicode == 65535:
+        dirname = os.path.join(dirname, python_version+'m')
+    # encode with ucs4
+    else:
+        dirname = os.path.join(dirname, python_version+'mu')
+elif 'macosx' in platform:
+    platform = 'macosx'
```
Comment (jamalex) [inline kolibri/utils/cli.py]: Some of the logic added to this file looks like it duplicates logic above -- can we combine/reuse?

File: kolibri/utils/cli.py [hunk 4]
```diff
@@ -14,10 +15,44 @@
...
+dirname = os.path.join(dirname, platform)
+
+# Check if the path is valid
+if os.path.exists(os.path.realpath(str(dirname))) == False:
+    logging.warning('Package cryptography is not supported in this platform.\n')
```
Comment (jamalex) [inline kolibri/utils/cli.py]: `cryptography` is an example of a package we'd use this for, but I don't think it makes sense mentioning it specifically here -- perhaps just "No C Extensions available for this platform"

File: kolibri/utils/cli.py (hunk 3 above)
Comment (lyw07) [reply kolibri/utils/cli.py]: Sorry do you mean the logic here looks like a duplicate of the logic in install_cext.py? So the logic in install_cext.py is to parse the whl files in pypi repositories and distribute them into different local directories depending on different platforms. The logic in this file is to get the platform of the running system and to find whether the corresponding c extension exists in local. In this case, I don't feel like we can combine these two. But I might misunderstand what you said, could you elaborate more on the duplicate logic part? Thank you!

Comment (jamalex) [reply kolibri/utils/cli.py]: I was thinking there was more overlap, and so we could pull some of it into a reusable function, but it looks like there are lots of little differences between them, so it probably wouldn't be a good idea, yeah.

**Comment by jamalex** [comment]
A general question (also for @aronasorman):

Ideally, we can keep the number of copies of each package to a minimum. For some packages, we need a different version for Python 2 vs 3. For other packages (like cryptography), what matters most is the architecture (and maybe the OS?). Is there some way for us to detect or hardcode what dimensions matter for each package, if it's indeed the case that, for example, for cryptography, we wouldn't need both:
`cryptography-2.0.3-cp35-cp35m-manylinux1_x86_64.whl`
and
`cryptography-2.0.3-cp36-cp36m-manylinux1_x86_64.whl`?

It looks like the md5 for those two files is actually different, however, so perhaps there's some difference, but not sure whether it's an important difference.

Thoughts?

**Comment by lyw07** [comment]
@jamalex Do you mean that we integrate wheel files for python3.5 and python3.6 on linux system into one wheel file that can work for both python3.5 and 3.6 environments? Thank you!

File: .gitignore [hunk 5]
```diff
@@ -9,6 +9,7 @@
 *.egg
 *.egg-info
 dist
+dist_cext
```
Comment (benjaoming) [inline .gitignore]: Hi @lyw07 !

Wondering.. Why does it have its own top-level `kolibri/dist_cext` instead of living in `kolibri/dist/cext/` ?

It would be nice to keep the folder structure clean - if possible - such that devs can easily grasp it. For reference, there has been an attempt to describe the structure here: http://kolibri.readthedocs.io/en/master/dev/files.html

It exemplifies both that it's hard to document, but essentially that it's hard for us to understand what's going on by shear intuition :)

File: .gitignore [hunk 6]
```diff
@@ -34,6 +36,7 @@ kolibrihome_test/
 
 # Translations
 kolibri/locale/en/
+kolibri/locale/**/django.mo
```
Comment (benjaoming) [inline .gitignore]: This seems unrelated?

File: kolibri/utils/cli.py [hunk 7]
```diff
@@ -14,10 +15,44 @@
...
+from kolibri import dist_cext as kolibri_dist_cext
+
 sys.path = sys.path + [
     os.path.realpath(os.path.dirname(kolibri_dist.__file__))
 ]
 
+
+"""
+Get the directory of dist_cext.
+"""
```
Comment (benjaoming) [inline kolibri/utils/cli.py]: Seeing that you have written a docstring for a function, I would suggest moving all this to a function to keep the module more readable.

File: .gitignore (hunk 5 above)
Comment (lyw07) [reply .gitignore]: Thank you for your suggestion. I will move dist_cext to dist/cext and update my PR soon!

File: .gitignore (hunk 6 above)
Comment (lyw07) [inline .gitignore]: Yes I'm not pretty sure why these files were generated so I git-ignored them. Do you have any suggestions where I should put this line? Or do you think I can just add it to the repository? Thank you!

File: kolibri/utils/cli.py (hunk 7 above)
Comment (lyw07) [reply kolibri/utils/cli.py]: Thanks! I'll put them into a function.

**Comment by benjaoming** [comment]
Woah. This is quite a serious change in architecture.

I would like to encourage that we move towards such changes with caution - it's a great idea to have optional compiled libraries, but it can seriously disrupt the downstream distributions of Kolibri.

One thing that I would like to suggest is to implement this with much less impact on the existing codebase. Not that I want to protect it from some great evil, but simply because Kolibri could *potentially* continue to be pure-python, while the compiled optimizations reside externally as a pluggable feature.

```
try:
     import quick_compiled_library
except ImportError:
    # fallback
```

On Linux-based systems, cryptography is already packaged. So it's possibly better to rely on this, since it's an optional improvement. For instance, that would give you [ARM packages for Raspberry Pi](http://archive.raspbian.org/raspbian/pool/main/p/python-cryptography/).

File: install_cexts.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+        imp.find_module('bs4')
+        print ('bs4 module exists')
+    except:
+        subprocess.call(['pip', 'install', 'bs4'])
+
+
+"""
+Distribute packages into folders according to the platform.
+"""
+def get_path_with_arch(platform, path):
```
Comment (benjaoming) [inline install_cexts.py]: If you move the docstrings into the function body, Sphinx will pick it up. Like so:

```
def get_path_with_arch(platform, path):
    """
    Distribute packages into folders according to the platform.
    """
```

File: install_cexts.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/env python
+import os
+import subprocess
+import argparse
+import imp
+
+
+DIST_CEXT = 'kolibri/dist/cext'
```
Comment (benjaoming) [inline install_cexts.py]: I came to wonder a bit after making my comments: Why should the compiled libraries sit in another target directory? Might they not be together with everything else in `kolibri/dist`?

As per my other comment, I'm sort of leaning towards having the compiled+optional dependencies distributed outside of Kolibri, so it's totally cool to leave it like this.

File: install_cexts.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+
+
+"""
+Check if requests and bs4 modules are installed.
+"""
+def _install_requests_and_bs4_if_needed():
+    try:
+        imp.find_module('requests')
+        print ('requests module exists')
+    except:
```
Comment (benjaoming) [inline install_cexts.py]: I'm nitpicking here, but having `except` catching ALL exceptions is an anti-pattern always - it's better to write `except ImportError` or whatever specific exception you want to catch. Small thing here, but good habit!

**Comment by lyw07** [comment]
Having a general error when using python3 to run the pex file generated by buildkite. 

`(kolibri) Lingyi-4:Downloads lingyiwang$ python3 kolibri-0.6.dev020170821165952-git.pex
Traceback (most recent call last):
  File "/usr/local/Cellar/python3/3.6.2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/Cellar/python3/3.6.2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "kolibri-0.6.dev020170821165952-git.pex/__main__.py", line 25, in <module>
  File "/Users/lingyiwang/Downloads/kolibri-0.6.dev020170821165952-git.pex/.bootstrap/_pex/pex_bootstrapper.py", line 94, in bootstrap_pex
  File "/Users/lingyiwang/Downloads/kolibri-0.6.dev020170821165952-git.pex/.bootstrap/_pex/finders.py", line 218, in register_finders
AttributeError: module 'importlib._bootstrap' has no attribute 'FileFinder'`

This error occurs when running not only the pex file for my PR but also other pex files in other PRs.
 
Mac os x 10.11.6
Python: 3.6.2 (installed from Homebrew 1.3.1)
pip: 9.0.1
node: v6.11.2
yarn: 0.27.5

**Comment by benjaoming** [comment]
> Having a general error when using python3 to run the pex file generated by buildkite.

Did you also build the pex file with python3?

**Comment by lyw07** [comment]
I built the pex files with python3.6 on my laptop (mac os), and it runs fine with Python 2 and 3 on my laptop and a ubuntu VM I set up on vagrant. But I haven't tried to build the pex file with python3.6 on buildkite yet.

**Comment by lyw07** [comment]
> Having a general error when using python3 to run the pex file generated by buildkite.

Problem fixed by creating pex file inside the build_whl.sh instead of using .buildkite/build_pex.sh to generate pex file. It is because .buildkite/build_pex.sh generates the pex file inside the actual host, where pex version is only 1.1.2, instead of inside a docker container. 
After this fix, pex file downloaded from buildkite runs successfully with Python3.5 and Python3.6 as well.

**Comment by benjaoming** [comment]
I would like to add another point regarding having C extensions in the general repo: For the test matrix, we need to expand by an extra dimension... a factor 2x on our build times. Everything has to be tested both with and without the C extensions.

This extra dimension would probably reside better in a separate repo?

File: kolibri/utils/cli.py [hunk 11]
```diff
@@ -139,15 +138,17 @@ def get_cext_path(dist_path):
...
 
 # Add path for c extensions to sys.path
-get_cext_path(os.path.dirname(kolibri_dist.__file__))
+get_cext_path(sys.path[-1])
+try:
+    import cryptography
+except ImportError:
+    # Fallback
+    logging.warning('No C Extensions available for this platform.\n')
+    sys.path = sys.path[:-1]
```
Comment (benjaoming) [inline kolibri/utils/cli.py]: This wouldn't be necessary if the C compiled modules shared the same directory `kolibri/dist` as the other distributed modules. I think from this example that it's probably better to use the same target package directory, you will be able to remove a lot of code.

Comment (lyw07) [reply kolibri/utils/cli.py]: Thanks! But I'm a bit confused here about putting them all into `kolibri/dist`. Since there are cryptography modules on mac/linux/windows, pip will complain that the module already exists if I install cryptography for linux in `kolibri/dist` after installing cryptography for mac in `kolibri/dist`. Then I guess I still need to install them into separate directories under `kolibri/dist` based on OS to avoid duplicate install in the same directory and insert the path of cryptography for specific OS into `sys.path`. And in that case, it is the same as the code here. So I'm not sure if I understand your suggestion correctly. Could you please give me a bit more explanations about your suggestion? Thank you @benjaoming !

File: kolibri/utils/cli.py [hunk 12]
```diff
@@ -145,6 +145,7 @@ def get_cext_path(dist_path):
 get_cext_path(sys.path[-1])
 try:
     import cryptography
+    version = cryptography.__version__ # get rid of Travis complain
```
Comment (benjaoming) [inline kolibri/utils/cli.py]: Is this travis complaining or the linting? You can use `  # noqa` for code lines that don't lint but aren't supposed to.

Comment (lyw07) [reply kolibri/utils/cli.py]: Thanks! Will do it now

File: kolibri/utils/cli.py [hunk 13]
```diff
@@ -144,7 +144,7 @@ def get_cext_path(dist_path):
 # Add path for c extensions to sys.path
 get_cext_path(sys.path[-1])
 try:
-    import cryptography # noqa
+    import cryptography  # noqa
```
Comment (benjaoming) [inline kolibri/utils/cli.py]: You can also use this pattern to avoid the extra commit:

```
# edit ./file
git add file
git commit - "Adding something"
git push origin branch-name
# edit ./file
git add file
git commit --amend  # Add to previous commit
git push origin branch-name -f  # Force-push to your own feature branch, that's okay
```

Comment (lyw07) [reply kolibri/utils/cli.py]: Thank you! That really helps

**Comment by benjaoming** [comment]
Trying to understand the scope of this better in general: Where is `cryptopgrahy` needed? Only in Morango? Are there other compiled dependencies? @jamalex 

I also understand from @aronasorman that he wants to start compiled optimizations in Iceqube.

We should discuss approaches to this: Do we want to put the multi-platform arch (multi-arch) in the source repos? Do we want to release source dists of Kolibri for multi-arch? Or do we use auto-detected optional dependencies?

**Comment by aronasorman** [comment]
From our C extensions discussion today, we decided to get this merged in, while thinking about how to move this to another repo, with the goal of creating a tool that will be useful in general to the open source community.

File: .buildkite/build_whl.sh [hunk 14]
```diff
@@ -6,3 +6,4 @@ make dockerenvdist
 buildkite-agent artifact upload 'dist/*.whl'
 buildkite-agent artifact upload 'dist/*.zip'
 buildkite-agent artifact upload 'dist/*.tar.gz'
+buildkite-agent artifact upload 'dist/*.pex'
```
Comment (aronasorman) [inline .buildkite/build_whl.sh:9]: This is a code smell, but I project that there'll be a refactoring in develop soon once `bake` is integrated, as that will require a refactoring of the Dockerfile.

File: windows_installer_docker_build/Dockerfile [hunk 15]
```diff
@@ -2,10 +2,6 @@ FROM ubuntu:xenial
 
 RUN apt-get -y update 
 
-# Install tff-mscorefonts so it won't prompt to accept the license later
-RUN echo ttf-mscorefonts-installer msttcorefonts/accepted-mscorefonts-eula select true | debconf-set-selections
-RUN apt-get install -y ttf-mscorefonts-installer
```
Comment (benjaoming) [inline windows_installer_docker_build/Dockerfile]: These are license prohibited anyways

Comment (benjaoming) [reply windows_installer_docker_build/Dockerfile]: `RUN echo ttf-mscorefonts-installer msttcorefonts/accepted-mscorefonts-eula select true` this part will accept a EULA on behalf of a user, essentially - I would think - putting the blame of license violations on LE.

Comment (lyw07) [reply windows_installer_docker_build/Dockerfile]: Sorry this was for testing the buildkite-agent virtual image I'm working on currently. I've already removed them :)

**Comment by lyw07** [comment]
@benjaoming Hi Ben! I've managed to squash the commits into three commits now. Do you think I need to continue to squash them into a single commit? Thank you!!

**Comment by benjaoming** [comment]
@lyw07 this is above average (or below to be precise) - of course the merge commits are nice to keep out, but let's be pragmatic (productive) about it too :)

**Comment by benjaoming** [comment]
Before we merge this and leave it to linger, can we have a clear plan of where to after?

Did we agree to move all of this to a separate repo?

We have a kind of naming convention of `kolibri-installers-<platform>`, I could imagine a repo called `kolibri-installers-pex` which builds Pex files for unified or individual platforms. It would be nice perhaps to open up this repository already now and identify some issues that we want to target once moving the logic/functionality from this PR:

1. Get rid of the `pip` hacks, we shouldn't be running some modified fork of pip, that's going to get heavy on us quickly (who's idea was it to fork and modify pip btw?)
1. Figure out the strategy and intention of the Pex builds.
1. Which platforms should be targeted?

**Comment by indirectlylit** [comment]
Is this work necessary for the upcoming 0.6?

**Comment by aronasorman** [comment]
> Get rid of the pip hacks, we shouldn't be running some modified fork of pip, that's going to get heavy on us quickly (who's idea was it to fork and modify pip btw?)

it was my idea -- pip doesn't download any source dists when you specify a `--platform` option that's different from the host platform. One of `cryptography`'s dependencies was a pure-python C parser, that was only distributed as a source dist. The main edit in the fork is to remove the part that raises that error.

I think, before we open any repo, let's have a discussion as a team first on what the next steps are. I don't think we settled on any one particular action plan yet.

**Comment by benjaoming** [comment]
@indirectlylit to my understanding, the idea of the Pex distribution with built-in compiled libraries might be nice to address the gap between 0.6 release w/o reliable installers.

If we don't utilize it for 0.6, I'm afraid that a lot of its purpose will be lost once we have installers.

@lyw07 I've created the repository:

https://github.com/learningequality/kolibri-installer-pex

**Comment by benjaoming** [comment]
@aronasorman Well, maybe we didn't settle on it? We discussed having the Pex build logic in a separate repo during the meeting, I was under the impression that we shared the idea of doing it in the future. IMO, we need to be quite alert about allowing installer/platform specific code to sneak into the source dist - it will always seem like the comfy or easy choice at first, but in the long run, it'll generate significant complexity/maintenance issues.. even if we opt for mono-repo stuff, it doesn't make the problem go away, just relocates the issues to a folder/file-specific domain, rather than repo-specific.
